engine/pom.xml

```diff
@@ -322,6 +322,51 @@
                     <skip>false</skip>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.moditect</groupId>
+                <artifactId>moditect-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>add-module-infos</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>add-module-info</goal>
+                        </goals>
+                        <configuration>
+                            <module>
+                                <moduleInfo>
+                                    <name>${java.module.name}</name>
+                                    <requires>
+                                        <!-- jboss annotations are getting pulled even though they are for compilation only -->
+                                        !jboss.logging.annotations;
+                                        *;
+                                    </requires>
+                                    <exports>
+                                        org.hibernate.validator.*.logging to org.jboss.logging, org.hibernate.validator.cdi;
+                                        org.hibernate.validator.internal to org.hibernate.validator.cdi;
+                                        org.hibernate.validator.internal.* to org.hibernate.validator.cdi;
+                                        org.hibernate.validator.*.internal.* to org.hibernate.validator.cdi;
+                                        *;
+                                    </exports>
+                                    <opens>
+                                        <!-- We open constraints for reflection as these can be instantiated by a CDI container -->
+                                        org.hibernate.validator.internal.constraintvalidators.bv;
+                                        org.hibernate.validator.internal.constraintvalidators.bv.*;
+                                        org.hibernate.validator.internal.constraintvalidators.hv;
+                                        org.hibernate.validator.internal.constraintvalidators.hv.*;
+                                        <!-- As both Validator and ValidatorFactory have to be accessible by CDI ... -->
+                                        org.hibernate.validator.internal.engine;
```

Comment:
You mean Weld accesses the Validator/ValidatorFactory impl? Why... ? How?

Reply:
```
Caused by: org.jboss.weld.exceptions.WeldException: WELD-001524: Unable to load proxy class for bean ValidatorFactoryBean [id=org.hibernate.validator.cdi.internal.ValidatorFactoryBean_default] with class class org.hibernate.validator.internal.engine.ValidatorFactoryImpl
        at org.jboss.weld.bean.proxy.ProxyFactory.getProxyClass(ProxyFactory.java:409)
        at org.jboss.weld.bean.proxy.ProxyFactory.run(ProxyFactory.java:360)
        at org.jboss.weld.bean.proxy.ProxyFactory.create(ProxyFactory.java:352)
        at org.jboss.weld.bean.proxy.ClientProxyFactory.create(ClientProxyFactory.java:83)
        at org.jboss.weld.bean.proxy.ClientProxyProvider.createClientProxy(ClientProxyProvider.java:205)
        at org.jboss.weld.bean.proxy.ClientProxyProvider.createClientProxy(ClientProxyProvider.java:195)
        at org.jboss.weld.bean.proxy.ClientProxyProvider$CreateClientProxy.apply(ClientProxyProvider.java:52)
        at org.jboss.weld.bean.proxy.ClientProxyProvider$CreateClientProxy.apply(ClientProxyProvider.java:48)
        at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.lambda$new$0(ReentrantMapBackedComputingCache.java:55)
        at org.jboss.weld.util.LazyValueHolder$1.computeValue(LazyValueHolder.java:32)
        at org.jboss.weld.util.LazyValueHolder.get(LazyValueHolder.java:46)
        at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.getValue(ReentrantMapBackedComputingCache.java:72)
        at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.getCastValue(ReentrantMapBackedComputingCache.java:78)
        at org.jboss.weld.bean.proxy.ClientProxyProvider.getClientProxy(ClientProxyProvider.java:229)
        at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:678)
        at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:666)
        at org.jboss.weld.manager.BeanManagerImpl.getInjectableReference(BeanManagerImpl.java:784)
        at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
        at org.jboss.weld.util.Beans.injectBoundFields(Beans.java:345)
        at org.jboss.weld.util.Beans.injectFieldsAndInitializers(Beans.java:356)
        at org.jboss.weld.injection.producer.ResourceInjector$1.proceed(ResourceInjector.java:69)
        at org.jboss.weld.injection.InjectionContextImpl.run(InjectionContextImpl.java:48)
        at org.jboss.weld.injection.producer.ResourceInjector.inject(ResourceInjector.java:71)
        at org.jboss.weld.injection.producer.BasicInjectionTarget.inject(BasicInjectionTarget.java:117)
        at org.jboss.arquillian.testenricher.cdi.CDIInjectionEnricher.injectNonContextualInstance(CDIInjectionEnricher.java:127)
        at org.jboss.arquillian.testenricher.cdi.CDIInjectionEnricher.injectClass(CDIInjectionEnricher.java:110)
        ... 68 more
Caused by: java.lang.RuntimeException: java.lang.IllegalAccessException: module org.hibernate.validator does not open org.hibernate.validator.internal.engine to unnamed module @6b81ce95
        at org.jboss.weld.bean.proxy.util.WeldDefaultProxyServices.defineWithMethodLookup(WeldDefaultProxyServices.java:151)
        at org.jboss.weld.bean.proxy.util.WeldDefaultProxyServices.defineClass(WeldDefaultProxyServices.java:60)
        at org.jboss.weld.bean.proxy.ProxyFactory.toClass(ProxyFactory.java:940)
        at org.jboss.weld.bean.proxy.ProxyFactory.createProxyClass(ProxyFactory.java:495)
        at org.jboss.weld.bean.proxy.ProxyFactory.getProxyClass(ProxyFactory.java:401)
        ... 93 more
Caused by: java.lang.IllegalAccessException: module org.hibernate.validator does not open org.hibernate.validator.internal.engine to unnamed module @6b81ce95
        at java.base/java.lang.invoke.MethodHandles.privateLookupIn(MethodHandles.java:287)
        at org.jboss.weld.bean.proxy.util.WeldDefaultProxyServices.defineWithMethodLookup(WeldDefaultProxyServices.java:148)
        ... 97 more
```

Reply:
It seems to me that the only reason CDI would need to access `ValidatorFactoryImpl` would be to instantiate it. But we instantiate the class ourselves in `org.hibernate.validator.cdi.internal.ValidatorFactoryBean#create`, so I think it's only being instantiated is for CDI proxies.

I don't know Weld very well, but in Arc (Quarkus's CDI impl) the CDI proxy would typically extend `ValidatorFactory`, not `ValidatorFactoryImpl`.

I see that here, we essentially expose the `ValidatorFactoryImpl` type in CDI:

https://github.com/hibernate/hibernate-validator/blob/f5715b41eee327a4442c072a2f513ac438314971/cdi/src/main/java/org/hibernate/validator/cdi/internal/ValidatorFactoryBean.java#L70-L74

IMO we shouldn't. The bean types should only include public types, so `org.hibernate.validator.HibernateValidatorFactory` and superinterfaces. Maybe `PredefinedScopeHibernateValidatorFactory` and superinterfaces, when relevant. But definitely not `ValidatorFactoryImpl`.

Maybe if you the exposed types of the CDI bean, the problem will be different?

For constraint validators, however, I'm not sure it's the same problem... I find it weird that Weld tries to enable constructor injection on these, though; we don't rely on CDI anywhere in our constraint validators, do we?

Reply:
> Maybe if you the exposed types of the CDI bean, the problem will be different?

Thanks for the suggestion, I'll give that a try! 😃 

> For constraint validators, however, I'm not sure it's the same problem... I find it weird that Weld tries to enable constructor injection on these, though; we don't rely on CDI anywhere in our constraint validators, do we?

hmm, right so we do not rely on CDI in any built-in constraints. There is this option for custom constraints if someone decides to do that (https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#example-constraint-validator-injected-bean) sooooooo maybe we update the factory here:

https://github.com/hibernate/hibernate-validator/blob/125feaa30e91fd149d29111bec5c3179b9f6feaf/cdi/src/main/java/org/hibernate/validator/cdi/internal/InjectingConstraintValidatorFactory.java#L40-L44

And if it is a built-in constraint, we return an instance and do not go through CDI ?

Reply:
> And if it is a built-in constraint, we return an instance and do not go through CDI ?

That looks like a plan. Now how you'll detect it's a built-in constraint is going to be interesting... Please remember some demos use the package name `org.hibernate.validator.demo` or something like that... Ideally you'd re-use some pre-existing code that's designed to detect built-in classes, if that exists.